feat: Langfuse integration, system-wide tracing configuration (LangSmith&Langfuse) (#253)
RobotecAI/rai: docs/tracing.md

# Tracing Configuration

> RAI supports tracing capabilities to help monitor and analyze the performance of your AI models. This document outlines how to configure tracing for your RAI project.

Reviewer: ```suggestion
RAI supports tracing capabilities to help monitor and analyze the performance of your AI models, at a minor performance cost. By default, tracing is off. This document outlines how to configure tracing for your RAI project.
```

## Configuration

> Tracing configuration is managed through the `config.toml` file. The relevant section for tracing is as follows:

Reviewer: ```suggestion
Tracing configuration is managed through the `config.toml` file. The relevant parameters for tracing are:
```

### Project Name

The `project` field under the `[tracing]` section sets the name for your tracing project. This name will be used to identify your project in the tracing tools.

> [!NOTE]  

> > Only LangSmith project name is currently used. Langfuse will upload traces to the default project.

Reviewer: ```suggestion
> Project name is currently only used by LangSmith. Langfuse will upload traces to the default project.
```

### LangSmith

[LangSmith](https://www.langchain.com/langsmith) is a platform for debugging, testing, and monitoring LangChain applications.

To enable LangSmith tracing:

1. Set `use_langsmith = true` in the `config.toml` file.
2. Set the `LANGCHAIN_API_KEY` environment variable with your LangSmith API key.

When enabled, LangSmith tracing will automatically be set up with the following environment variables:

> ### Langfuse

Reviewer: I would always start with the open source option when reading from the top, and make the distinction extra clear, i.e.:

### Langfuse (open-source)
### LangSmith (closed-source, paid)

[Langfuse](https://langfuse.com/) is an open-source observability & analytics platform for LLM applications.

To enable Langfuse tracing:

1. Set `use_langfuse = true` in the `config.toml` file.
2. Set the `LANGFUSE_PUBLIC_KEY` and `LANGFUSE_SECRET_KEY` environment variables with your Langfuse credentials.
3. Optionally, you can specify a custom Langfuse host by modifying the `host` field under `[tracing.langfuse]`.

## Usage

The tracing configuration is automatically loaded and applied when initializing models in RAI. The `get_tracing_callbacks()` function in `model_initialization.py` sets up the appropriate callbacks based on your configuration:

This function returns a list of callback handlers that can be used with LangChain.

## Troubleshooting

If you encounter issues with tracing:

1. Ensure all required environment variables are set correctly.

> 2. Check that the `use_langsmith` and `use_langfuse` flags are set to `true` in `config.toml` if you intend to use these services.

Reviewer: ```suggestion
2. Check whether tracing is on by checking whether `use_langsmith` or `use_langfuse` flag is set to `true` in `config.toml`.
```
